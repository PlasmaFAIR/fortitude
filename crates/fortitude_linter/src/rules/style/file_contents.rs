use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for multiple modules in one file
///
/// ## Why is this bad?
/// Placing each module into its own file improves maintainability
/// by making each module easier to locate for developers, and also
/// making dependency generation in build systems easier.
#[derive(ViolationMetadata)]
pub(crate) struct MultipleModules {}

impl Violation for MultipleModules {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Multiple modules in one file, split into one module per file".to_string()
    }
}

impl AstRule for MultipleModules {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let violations: Vec<Diagnostic> = node
            .children(&mut node.walk())
            .filter(|node| node.kind() == "module")
            .skip(1)
            .map(|m| -> Diagnostic {
                let m_first = m.child(0).unwrap_or(m);
                Diagnostic::from_node(MultipleModules {}, &m_first)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["translation_unit"]
    }
}

/// ## What it does
/// Checks for programs and modules in one file
///
/// ## Why is this bad?
/// Separating top-level constructs into their own files improves
/// maintainability by making each easier to locate for developers,
/// and also making dependency generation in build systems easier.
#[derive(ViolationMetadata)]
pub(crate) struct ProgramWithModule {}

impl Violation for ProgramWithModule {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Program and module in one file, split into their own files".to_string()
    }
}

impl AstRule for ProgramWithModule {
    fn check(_settings: &Settings, node: &Node, _src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // There must be a program statement to trigger this rule
        if !node
            .children(&mut node.walk())
            .any(|node| node.kind() == "program")
        {
            return None;
        }

        // Mark the violation on the second, and subsequent, occurrences
        let violations: Vec<Diagnostic> = node
            .children(&mut node.walk())
            .filter(|node| node.kind() == "module" || node.kind() == "program")
            .skip(1)
            .map(|m| -> Diagnostic {
                let m_first = m.child(0).unwrap_or(m);
                Diagnostic::from_node(ProgramWithModule {}, &m_first)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["translation_unit"]
    }
}
