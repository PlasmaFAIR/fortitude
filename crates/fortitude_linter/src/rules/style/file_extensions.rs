use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_text_size::TextRange;

use crate::settings::Settings;
use crate::PathRule;
use std::path::Path;

/// ## What it does
/// Checks for use of standard file extensions.
///
/// ## Why is it bad?
/// The standard file extensions for modern (free-form) Fortran are '.f90' or  '.F90'.
/// Forms that reference later Fortran standards such as '.f08' or '.F95' may be rejected
/// by some compilers and build tools.
#[derive(ViolationMetadata)]
pub(crate) struct NonStandardFileExtension {}

impl Violation for NonStandardFileExtension {
    #[derive_message_formats]
    fn message(&self) -> String {
        "file extension should be '.f90' or '.F90'".to_string()
    }
}

impl PathRule for NonStandardFileExtension {
    fn check(_settings: &Settings, path: &Path) -> Option<Diagnostic> {
        match path.extension() {
            Some(ext) => {
                // Must check like this as ext is an OsStr
                if ["f90", "F90"].iter().any(|&x| x == ext) {
                    None
                } else {
                    Some(Diagnostic::new(Self {}, TextRange::default()))
                }
            }
            None => Some(Diagnostic::new(Self {}, TextRange::default())),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bad_file_extension() {
        let path = Path::new("my/dir/to/file.f95");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path),
            Some(Diagnostic::new(
                NonStandardFileExtension {},
                TextRange::default()
            )),
        );
    }

    #[test]
    fn test_missing_file_extension() {
        let path = Path::new("my/dir/to/file");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path),
            Some(Diagnostic::new(
                NonStandardFileExtension {},
                TextRange::default()
            )),
        );
    }

    #[test]
    fn test_correct_file_extensions() {
        let path1 = Path::new("my/dir/to/file.f90");
        let path2 = Path::new("my/dir/to/file.F90");
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path1),
            None
        );
        assert_eq!(
            NonStandardFileExtension::check(&Settings::default(), path2),
            None
        );
    }
}
