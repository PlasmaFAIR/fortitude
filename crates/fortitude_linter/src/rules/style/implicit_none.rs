/// Defines rules that raise errors if implicit typing is in use.
use crate::ast::FortitudeNode;
use crate::rules::correctness::implicit_typing::{
    child_is_implicit_none, implicit_statement_is_none,
};
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for unnecessary `implicit none` in module procedures
///
/// ## Why is this bad?
/// If a module has 'implicit none' set, it is not necessary to set it in contained
/// functions and subroutines (except when using interfaces).
#[derive(ViolationMetadata)]
pub struct SuperfluousImplicitNone {
    entity: String,
}

impl AlwaysFixableViolation for SuperfluousImplicitNone {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { entity } = self;
        format!("'implicit none' set on the enclosing {entity}")
    }

    fn fix_title(&self) -> String {
        "Remove unnecessary 'implicit none'".to_string()
    }
}

impl AstRule for SuperfluousImplicitNone {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if !implicit_statement_is_none(node) {
            return None;
        }
        let parent = node.parent()?;
        if matches!(parent.kind(), "function" | "subroutine") {
            for ancestor in parent.ancestors() {
                let kind = ancestor.kind();
                match kind {
                    "module" | "submodule" | "program" | "function" | "subroutine" => {
                        if !child_is_implicit_none(&ancestor) {
                            continue;
                        }
                        let entity = kind.to_string();
                        let fix = Fix::safe_edit(node.edit_delete(src));
                        return some_vec![
                            Diagnostic::from_node(Self { entity }, node).with_fix(fix)
                        ];
                    }
                    "interface" => {
                        break;
                    }
                    _ => {
                        continue;
                    }
                }
            }
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["implicit_statement"]
    }
}
