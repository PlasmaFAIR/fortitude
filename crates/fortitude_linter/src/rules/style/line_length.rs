/// Defines rules that govern line length.
use crate::settings::Settings;
use crate::TextRule;
use lazy_regex::regex_is_match;
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_source_file::UniversalNewlines;
use ruff_text_size::{TextLen, TextRange, TextSize};

/// ## What does it do?
/// Checks line length isn't too long
///
/// ## Why is this bad?
/// Long lines are more difficult to read, and may not fit on some developers'
/// terminals. The line continuation character '&' may be used to split a long line
/// across multiple lines, and overly long expressions may be broken down into
/// multiple parts.
///
/// The maximum line length can be changed using the flag `--line-length=N`. The
/// default maximum line length is 100 characters. This is a fair bit more than the
/// traditional 80, but due to the verbosity of modern Fortran it can sometimes be
/// difficult to squeeze lines into that width, especially when using large indents
/// and multiple levels of indentation.
///
/// Some lines that are longer than the maximum length may be acceptable, such as
/// long strings or comments. This is to allow for long URLs or other text that cannot
/// be reasonably split across multiple lines.
///
/// Note that the Fortran standard states a maximum line length of 132 characters,
/// and while some modern compilers will support longer lines, for portability it
/// is recommended to stay beneath this limit.
#[derive(ViolationMetadata)]
pub(crate) struct LineTooLong {
    max_length: usize,
    actual_length: usize,
}

impl Violation for LineTooLong {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self {
            max_length,
            actual_length,
        } = self;
        format!("line length of {actual_length}, exceeds maximum {max_length}")
    }
}

impl TextRule for LineTooLong {
    fn check(settings: &Settings, source_file: &SourceFile) -> Vec<Diagnostic> {
        let source = source_file.to_source_code();
        let max_length = settings.check.line_length;
        let mut violations = Vec::new();
        for line in source.text().universal_newlines() {
            // Note: Can't use string.len(), as that gives byte length, not char length
            let actual_length = line.chars().count();
            if actual_length > max_length {
                // Are we ending on a string or comment? If so, we'll allow it through, as it may
                // contain something like a long URL that cannot be reasonably split across multiple
                // lines.
                if regex_is_match!(r#"(["']\w*&?$)|(!.*$)|(^\w*&)"#, line.as_str()) {
                    continue;
                }
                // Get the byte range from the first character that oversteps the limit
                // to the end of the line
                let extra_bytes: TextSize = line
                    .chars()
                    .rev()
                    .take(actual_length - max_length)
                    .map(TextLen::text_len)
                    .sum();
                let range = TextRange::new(line.end() - extra_bytes, line.end());
                violations.push(Diagnostic::new(
                    Self {
                        max_length,
                        actual_length,
                    },
                    range,
                ));
            }
        }
        violations
    }
}
