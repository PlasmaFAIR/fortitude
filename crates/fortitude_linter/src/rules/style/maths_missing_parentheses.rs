use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks if mathematical expressions are missing parentheses when operators have
/// different precedences.
///
/// ## Why is this bad?
/// Long or complex expressions can be difficult or confusing to read, especially when
/// mixing operators with different precedences. Adding parentheses can clarify the code
/// and make the author's intent clearer, reducing the likelihood of misunderstandings
/// or bugs.
///
/// ## Example
/// ```f90
/// x = 1. + 2. * 3. - 4. / 5.
/// ```
///
/// Use instead:
/// ```f90
/// x = 1. + (2. * 3.) - (4. / 5.)
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MathsMissingParentheses {
    op1: String,
    op2: String,
}

impl AlwaysFixableViolation for MathsMissingParentheses {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { op1, op2 } = &self;
        format!("'{op1}' has higher precedence than '{op2}'; add parentheses for clarity")
    }

    fn fix_title(&self) -> String {
        "Add parentheses".to_string()
    }
}

#[derive(PartialEq)]
enum Operator {
    Exponential,
    Multiplication,
    Division,
    Addition,
    Subtraction,
}

impl Operator {
    fn precedence(&self) -> u8 {
        use Operator::*;
        match self {
            Exponential => 4,
            Multiplication | Division => 3,
            Addition | Subtraction => 2,
        }
    }
}

impl TryFrom<&str> for Operator {
    type Error = &'static str;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        use Operator::*;
        match value {
            "**" => Ok(Exponential),
            "*" => Ok(Multiplication),
            "/" => Ok(Division),
            "-" => Ok(Addition),
            "+" => Ok(Subtraction),
            _ => Err("not a maths operator"),
        }
    }
}

impl From<Operator> for String {
    fn from(value: Operator) -> String {
        use Operator::*;
        match value {
            Exponential => "**",
            Multiplication => "*",
            Division => "/",
            Addition => "-",
            Subtraction => "+",
        }
        .to_string()
    }
}

impl AstRule for MathsMissingParentheses {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let text = src.source_text();
        let op: Operator = node.child(1)?.to_text(text)?.try_into().ok()?;

        let parent = node.parent()?;
        if !matches!(parent.kind(), "math_expression" | "unary_expression") {
            return None;
        }

        let edit = node.edit_replacement(src, format!("({})", node.to_text(text)?));
        let fix = Fix::safe_edit(edit);
        if op == Operator::Exponential && parent.kind() == "unary_expression" {
            let op2 = parent.child(0)?.to_text(text)?.to_string();
            let op1: String = op.into();
            return some_vec!(Diagnostic::from_node(Self { op1, op2 }, node).with_fix(fix));
        }

        let parent_op: Operator = parent.child(1)?.to_text(text)?.try_into().ok()?;

        if op.precedence() == parent_op.precedence() {
            return None;
        }

        let (op1, op2): (String, String) = if op.precedence() > parent_op.precedence() {
            (op.into(), parent_op.into())
        } else {
            (parent_op.into(), op.into())
        };

        some_vec!(Diagnostic::from_node(Self { op1, op2 }, node).with_fix(fix))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["math_expression"]
    }
}
