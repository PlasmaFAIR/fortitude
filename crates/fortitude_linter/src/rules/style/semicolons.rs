use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_text_size::TextSize;
use tree_sitter::Node;

use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};

fn semicolon_is_superfluous(node: &Node) -> bool {
    let line_number = node.start_position().row;
    let prev_node = node.prev_sibling();
    let next_node = node.next_sibling();
    // Test it has at least one sibling.
    if prev_node.is_none() && next_node.is_none() {
        return true;
    }
    // Test it is at beginning of a line. If the previous sibling is on an earlier line,
    // or if there is no previous sibling, then it is at the beginning of a line.
    // Also check that the previous sibling isn't a semicolon!
    if let Some(prev_node) = prev_node {
        let prev_line_number = prev_node.start_position().row;
        if prev_line_number < line_number {
            return true;
        }
        if prev_line_number == line_number && prev_node.kind() == ";" {
            return true;
        }
    } else {
        return true;
    }
    // Test it is at the end of a line. If the next sibling is on a later line, or if
    // there is no next sibling, then it is at the end of a line. Also check that the
    // next sibling isn't a comment or another semicolon!
    if let Some(next_node) = next_node {
        let next_line_number = next_node.start_position().row;
        if next_line_number > line_number {
            return true;
        }
        if next_line_number == line_number && next_node.kind() == "comment" {
            return true;
        }
    } else {
        return true;
    }
    false
}

/// ## What does it do?
/// Catches a semicolon at the end of a line of code.
///
/// ## Why is this bad?
/// Many languages use semicolons to denote the end of a statement, but in Fortran each
/// line of code is considered its own statement (unless it ends with a line
/// continuation character, `'&'`). Semicolons may be used to separate multiple
/// statements written on the same line, but a semicolon at the end of a line has no
/// effect.
///
/// A semicolon at the beginning of a statement similarly has no effect, nor do
/// multiple semicolons in sequence.
#[derive(ViolationMetadata)]
pub(crate) struct SuperfluousSemicolon {}

impl AlwaysFixableViolation for SuperfluousSemicolon {
    #[derive_message_formats]
    fn message(&self) -> String {
        "unnecessary semicolon".to_string()
    }

    fn fix_title(&self) -> String {
        "Remove this character".to_string()
    }
}

impl AstRule for SuperfluousSemicolon {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if semicolon_is_superfluous(node) {
            let edit = node.edit_delete(src);
            return some_vec!(Diagnostic::from_node(Self {}, node).with_fix(Fix::safe_edit(edit)));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![";"]
    }
}

/// ## What does it do?
/// Catches multiple statements on the same line separated by a semicolon.
///
/// ## Why is this bad?
/// This can have a detrimental effect on code readability.
#[derive(ViolationMetadata)]
pub(crate) struct MultipleStatementsPerLine {}

impl AlwaysFixableViolation for MultipleStatementsPerLine {
    #[derive_message_formats]
    fn message(&self) -> String {
        "multiple statements per line".to_string()
    }

    fn fix_title(&self) -> String {
        "Separate over two lines".to_string()
    }
}

impl AstRule for MultipleStatementsPerLine {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        if semicolon_is_superfluous(node) {
            return None;
        }
        let indentation = node.indentation(src);
        let start = node.start_byte();
        let mut end = node.end_byte();
        let text = src.source_text().as_bytes();
        while text[end] == b' ' || text[end] == b'\t' {
            end += 1;
        }
        let edit = Edit::replacement(
            format!("\n{indentation}"),
            TextSize::try_from(start).unwrap(),
            TextSize::try_from(end).unwrap(),
        );

        some_vec!(Diagnostic::from_node(Self {}, node).with_fix(Fix::safe_edit(edit)))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![";"]
    }
}
