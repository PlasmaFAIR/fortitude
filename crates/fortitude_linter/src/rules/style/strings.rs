use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use ruff_text_size::{TextLen, TextSize};
use settings::Quote;
use tree_sitter::Node;

use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};

/// ## What does it do?
/// Catches use of single- or double-quoted strings, depending on the value of
/// [`check.strings.quotes`] option.
///
/// ## Why is this bad?
/// For consistency, all strings should be either single-quoted or double-quoted.
/// Exceptions are made for strings containing escaped quotes.
///
/// ## Example
/// ```f90
/// foo = 'bar'
/// ```
///
/// Assuming `quotes` is set to `double`, use instead:
/// ```f90
/// foo = "bar"
/// ```
///
/// ## Options
/// - `check.strings.quotes`
#[derive(ViolationMetadata)]
pub(crate) struct BadQuoteString {
    preferred_quote: Quote,
    contains_escaped_quotes: bool,
}

impl Violation for BadQuoteString {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        match self.preferred_quote {
            Quote::Double => "String uses single quotes but double quotes preferred".to_string(),
            Quote::Single => "String uses double quotes but single quotes preferred".to_string(),
        }
    }

    fn fix_title(&self) -> Option<String> {
        if self.contains_escaped_quotes {
            return None;
        }
        let title = match self.preferred_quote {
            Quote::Double => "Replace single quotes with double quotes",
            Quote::Single => "Replace double quotes with single quotes",
        };
        Some(title.to_string())
    }
}

impl AstRule for BadQuoteString {
    fn check(settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let preferred_quote = settings.check.strings.quotes;
        let bad_quote = preferred_quote.opposite();

        let text = node.to_text(src.source_text())?;
        if text.starts_with(bad_quote.as_char())
            && text.ends_with(bad_quote.as_char())
            && !text.contains(preferred_quote.as_char())
        {
            // Search for occurrence of escaped single quotes within the string.
            // These are double single quotes, e.g. "''"
            if text.contains(bad_quote.escaped()) && text.len() > 2 {
                return Some(vec![Diagnostic::from_node(
                    Self {
                        preferred_quote,
                        contains_escaped_quotes: true,
                    },
                    node,
                )]);
            }

            let start_byte = node.start_textsize();
            let end_byte = node.end_textsize();
            let edit_start = Edit::replacement(
                preferred_quote.as_char().to_string(),
                start_byte,
                start_byte + TextSize::from(1),
            );
            let edit_end = Edit::replacement(
                preferred_quote.as_char().to_string(),
                end_byte - TextSize::from(1),
                end_byte,
            );
            return some_vec!(Diagnostic::from_node(
                Self {
                    preferred_quote,
                    contains_escaped_quotes: false,
                },
                node,
            )
            .with_fix(Fix::safe_edits(edit_start, [edit_end])));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["string_literal"]
    }
}

/// ## What it does
/// Checks for strings that include escaped quotes that can be removed if the
/// quote style is changed.
///
/// ## Why is this bad?
/// It's preferable to avoid escaped quotes in strings. By changing the
/// outer quote style, you can avoid escaping inner quotes.
///
/// ## Example
/// ```f90
/// foo = 'bar''s'
/// ```
///
/// Use instead:
/// ```f90
/// foo = "bar's"
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct AvoidableEscapedQuote;

impl AlwaysFixableViolation for AvoidableEscapedQuote {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Avoidable escaped quotes".to_string()
    }

    fn fix_title(&self) -> String {
        "Change outer quotes to avoid escaping inner quotes".to_string()
    }
}

impl AstRule for AvoidableEscapedQuote {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let text = node.to_text(src.source_text())?;
        if text.len() <= 2 {
            return None;
        }
        let quote_style = Quote::from_literal(node, text);

        if !text.contains(quote_style.escaped()) || text.contains(quote_style.opposite().as_char())
        {
            return None;
        }

        // Because the kind appears as a child node, the interesting
        // literal bit doesn't start at the beginning of the node
        let (start, kind) = if let Some(kind) = node.child_by_field_name("kind") {
            let kind_text = kind.to_text(src.source_text())?;
            (kind_text.len() + 1, format!("{kind_text}_"))
        } else {
            (0, "".to_string())
        };

        let end = text.text_len() - TextSize::new(1);
        let contents = &text[start + 1..end.to_usize()];
        let fixed = format!(
            "{kind}{quote}{value}{quote}",
            quote = quote_style.opposite().as_char(),
            value = unescape_string(contents, quote_style.as_char())
        );

        let edit = node.edit_replacement(src, fixed);
        // Offset start of node by kind, if any
        let range = node
            .textrange()
            .add_start(TextSize::try_from(start).unwrap());
        some_vec!(Diagnostic::new(Self, range).with_fix(Fix::safe_edit(edit)))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["string_literal"]
    }
}

fn unescape_string(haystack: &str, quote: char) -> String {
    let mut fixed_contents = String::with_capacity(haystack.len());

    let mut chars = haystack.chars().peekable();
    let mut seen_quote = false;
    while let Some(char_) = chars.next() {
        // Not a quote, or the previous character was a quote, which
        // we've removed
        if char_ != quote || seen_quote {
            fixed_contents.push(char_);
            seen_quote = false;
            continue;
        }
        // If we're at the end of the line
        let Some(next_char) = chars.peek() else {
            fixed_contents.push(char_);
            continue;
        };
        // Remove first of two consecutive quotes
        if *next_char == quote {
            seen_quote = true;
            continue;
        }
        fixed_contents.push(char_);
    }
    fixed_contents
}

pub mod settings {
    use super::*;
    use crate::display_settings;
    use ruff_macros::CacheKey;
    use serde::{Deserialize, Serialize};
    use std::fmt::{Display, Formatter};

    #[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize, CacheKey)]
    #[serde(deny_unknown_fields, rename_all = "kebab-case")]
    pub enum Quote {
        /// Use double quotes.
        Double,
        /// Use single quotes.
        Single,
    }

    impl Default for Quote {
        fn default() -> Self {
            Self::Double
        }
    }

    impl TryFrom<char> for Quote {
        type Error = &'static str;

        fn try_from(value: char) -> Result<Self, Self::Error> {
            match value {
                '"' => Ok(Self::Double),
                '\'' => Ok(Self::Single),
                _ => Err("not a quote"),
            }
        }
    }

    impl Quote {
        #[must_use]
        pub const fn opposite(self) -> Self {
            match self {
                Self::Double => Self::Single,
                Self::Single => Self::Double,
            }
        }

        #[must_use]
        pub const fn escaped(self) -> &'static str {
            match self {
                Self::Double => r#""""#,
                Self::Single => r#"''"#,
            }
        }

        /// Get the character used to represent this quote.
        pub const fn as_char(self) -> char {
            match self {
                Self::Double => '"',
                Self::Single => '\'',
            }
        }

        pub fn from_literal(node: &Node, text: &str) -> Self {
            let mut start = TextSize::new(0);
            if let Some(kind) = node.child_by_field_name("kind") {
                start = kind.textrange().len() + TextSize::new(1);
            }
            let first_quote = text
                .chars()
                .nth(start.to_usize())
                .expect("couldn't slice string literal correctly");
            Quote::try_from(first_quote).expect("string literal doesn't begin with a quote")
        }
    }

    impl Display for Quote {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            match self {
                Self::Double => write!(f, "double"),
                Self::Single => write!(f, "single"),
            }
        }
    }

    #[derive(Debug, Clone, Default, CacheKey)]
    pub struct Settings {
        pub quotes: Quote,
    }

    impl Display for Settings {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            display_settings! {
                formatter = f,
                namespace = "check.strings",
                fields = [self.quotes]
            }
            Ok(())
        }
    }
}
