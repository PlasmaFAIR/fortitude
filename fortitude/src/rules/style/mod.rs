pub mod double_colon_in_decl;
pub mod end_statements;
pub mod file_contents;
pub mod file_extensions;
pub mod functions;
pub mod implicit_none;
pub mod keywords;
pub mod line_length;
pub mod semicolons;
pub mod whitespace;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::style::keywords;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::LineTooLong, Path::new("S001.f90"))]
    #[test_case(Rule::UnnamedEndStatement, Path::new("S061.f90"))]
    #[test_case(Rule::MissingDoubleColon, Path::new("S071.f90"))]
    #[test_case(Rule::SuperfluousSemicolon, Path::new("S081.f90"))]
    #[test_case(Rule::MultipleStatementsPerLine, Path::new("S082.f90"))]
    #[test_case(Rule::TrailingWhitespace, Path::new("S101.f90"))]
    #[test_case(Rule::IncorrectSpaceBeforeComment, Path::new("S102.f90"))]
    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201.f90"))]
    #[test_case(Rule::MultipleModules, Path::new("S211.f90"))]
    #[test_case(Rule::ProgramWithModule, Path::new("S212.f90"))]
    #[test_case(Rule::FunctionMissingResult, Path::new("S221.f90"))]
    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::SuperfluousImplicitNone, Path::new("S201_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test_case(Rule::LineTooLong, Path::new("S001_line_length_20.f90"))]
    fn line_too_long_line_length_20(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                line_length: 20,
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::KeywordsMissingSpace, Path::new("S231.f90"))]
    #[test_case(Rule::KeywordHasWhitespace, Path::new("S231.f90"))]
    fn keyword_whitespace_include_inout_goto(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!(
            "{}_{}_include_inout_goto",
            rule_code.as_ref(),
            path.to_string_lossy()
        );

        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                keyword_whitespace: keywords::settings::Settings {
                    inout_with_space: true,
                    goto_with_space: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("style").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
