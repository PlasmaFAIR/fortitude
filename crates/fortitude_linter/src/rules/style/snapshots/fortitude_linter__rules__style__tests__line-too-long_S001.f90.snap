---
source: fortitude/src/rules/style/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/style/S001.f90:2:101: S001 line length of 222, exceeds maximum 100
  |
1 | ...
2 | ...ally_really_really_really_really_really_really_really_really_really_really_long_module_name, only : integer_working_precision
  |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ S001
3 | ...
4 | ...
  |
